source/extensions/quic_listeners/quiche/envoy_quic_server_stream.cc

```diff
@@ -14,50 +14,118 @@
 #include "quiche/quic/core/http/quic_header_list.h"
 #include "quiche/quic/core/quic_session.h"
 #include "quiche/spdy/core/spdy_header_block.h"
-
+#include "extensions/quic_listeners/quiche/platform/quic_mem_slice_span_impl.h"
 #pragma GCC diagnostic pop
 
 #include "extensions/quic_listeners/quiche/envoy_quic_utils.h"
+#include "extensions/quic_listeners/quiche/envoy_quic_server_session.h"
+
 #include "common/buffer/buffer_impl.h"
 #include "common/http/header_map_impl.h"
 #include "common/common/assert.h"
 
 namespace Envoy {
 namespace Quic {
 
+EnvoyQuicServerStream::EnvoyQuicServerStream(quic::QuicStreamId id, quic::QuicSpdySession* session,
+                                             quic::StreamType type)
+    : quic::QuicSpdyServerStreamBase(id, session, type),
+      EnvoyQuicStream(
+          2 * GetQuicFlag(FLAGS_quic_buffered_data_threshold),
+          [this]() { runLowWatermarkCallbacks(); }, [this]() { runHighWatermarkCallbacks(); }) {}
+
+EnvoyQuicServerStream::EnvoyQuicServerStream(quic::PendingStream* pending,
+                                             quic::QuicSpdySession* session, quic::StreamType type)
+    : quic::QuicSpdyServerStreamBase(pending, session, type),
+      EnvoyQuicStream(
+          session->config()->ReceivedInitialStreamFlowControlWindowBytes(),
+          [this]() { runLowWatermarkCallbacks(); }, [this]() { runHighWatermarkCallbacks(); }) {}
+
 void EnvoyQuicServerStream::encode100ContinueHeaders(const Http::HeaderMap& headers) {
   ASSERT(headers.Status()->value() == "100");
   encodeHeaders(headers, false);
 }
-void EnvoyQuicServerStream::encodeHeaders(const Http::HeaderMap& /*headers*/, bool /*end_stream*/) {
-  NOT_IMPLEMENTED_GCOVR_EXCL_LINE;
+
+void EnvoyQuicServerStream::encodeHeaders(const Http::HeaderMap& headers, bool end_stream) {
+  ENVOY_STREAM_LOG(debug, "encodeHeaders (end_stream={}) {}.", *this, end_stream, headers);
+  WriteHeaders(envoyHeadersToSpdyHeaderBlock(headers), end_stream, nullptr);
+  local_end_stream_ = end_stream;
 }
-void EnvoyQuicServerStream::encodeData(Buffer::Instance& /*data*/, bool /*end_stream*/) {
-  NOT_IMPLEMENTED_GCOVR_EXCL_LINE;
+
+void EnvoyQuicServerStream::encodeData(Buffer::Instance& data, bool end_stream) {
+  ENVOY_STREAM_LOG(debug, "encodeData (end_stream={}) of {} bytes.", *this, end_stream,
+                   data.length());
+  local_end_stream_ = end_stream;
+  // This is counting not serialized bytes in the send buffer.
+  uint64_t bytes_to_send_old = BufferedDataBytes();
+  // QUIC stream must take all.
+  WriteBodySlices(quic::QuicMemSliceSpan(quic::QuicMemSliceSpanImpl(data)), end_stream);
```

Comment:
I'm a bit confused about how this works.
I thought that WriteBodySlices consumed some amount of data, and returned how many bytes are consumed.  Where are the rest going?  Are they getting drained from data and then deleted as the transient QuicMemSlice bits go away or OnCanWrite?
I think this could benefit from some doccing up like in source/docs/flow_control.md talking about how things should work, and I can take another pass once we have it written up?

Reply:
Added docs/quiche_integration.md. PTAL

Reply:
It seems dangerous to ignore the return value of WriteBodySlices. Can we consider the follow behavior instead:
1. Use a very large buffered_data_threshold. This threshold has little to do with upstream flow control window etc, it is only used to prevent pathological conditions.
2. When we WriteBodySlices, we only consider it success if all data is consumed. Otherwise close the connection.
3. After a successful WriteBodySlices, run the watermark callbacks if needed.